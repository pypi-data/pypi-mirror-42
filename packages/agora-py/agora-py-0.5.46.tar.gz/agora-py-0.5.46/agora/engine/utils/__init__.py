"""
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Ontology Engineering Group
        http://www.oeg-upm.net/
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Copyright (C) 2016 Ontology Engineering Group.
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

            http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
"""
import os
from threading import Lock

__author__ = 'Fernando Serena'


class Singleton(type):
    _lock = Lock()

    def __call__(cls, *args, **kwargs):
        with cls._lock:
            id = kwargs.get('id', '')
            if id not in cls.instances:
                cls.instances[id] = super(Singleton, cls).__call__(*args, **kwargs)
            return cls.instances[id]


def prepare_store_path(base, path):
    if base and not os.path.exists(base):
        os.makedirs(base)
    if base or path:
        full_path = '/'.join(filter(lambda x: x, [base, path]))
        if not os.path.exists(full_path):
            os.makedirs(full_path)
        return full_path


class Semaphore(object):
    def __init__(self):
        self.value = -1

    def __enter__(self):
        self.value = 0

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.value = 1

    def isSet(self):
        return self.value == 1

    def set(self):
        self.value = 1


def get_immediate_subdirectories(a_dir):
    if not os.path.exists(a_dir):
        os.makedirs(a_dir)
    return [name for name in os.listdir(a_dir)
            if os.path.isdir(os.path.join(a_dir, name))]


class Wrapper(object):
    def __init__(self, object):
        self.__lock = Lock()
        self.__attr_locks = {}

        if isinstance(object, Wrapper):
            self.__object = object.__object
            self.__argret = object.__argret
            self.__attr_locks = object.__attr_locks
            self.__lock = object.__lock
        else:
            self.__object = object
            self.__argret = {}

    def decorate(self, f):
        def wrapper(*args, **kwargs):

            key = f.func_name + str(args)
            with self.__lock:
                if key not in self.__attr_locks:
                    self.__attr_locks[key] = Lock()

            with self.__attr_locks[key]:
                if key not in self.__argret:
                    ret = f(*args, **kwargs)

                    if kwargs:
                        return ret
                    if ret is not None:
                        self.__argret[key] = ret

                return self.__argret.get(key, None)

        return wrapper

    def __getattr__(self, item):

        if item.startswith('__'):
            return self.__object.__getattribute__(item)

        with self.__lock:
            if item not in self.__attr_locks:
                self.__attr_locks[item] = Lock()

        with self.__attr_locks[item]:
            if item not in self.__argret:
                attr = self.__object.__getattribute__(item)
                if hasattr(attr, '__call__'):
                    return self.decorate(attr)
                else:
                    self.__argret[item] = attr

        return self.__argret[item]


stopped = Semaphore()
