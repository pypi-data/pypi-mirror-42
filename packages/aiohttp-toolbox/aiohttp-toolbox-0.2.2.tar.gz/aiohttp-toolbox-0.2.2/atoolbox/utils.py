import json
import re
from typing import Any, Tuple, Type, TypeVar, Union

from aiohttp.web import Response
from aiohttp.web_exceptions import HTTPException
from pydantic import BaseModel, ValidationError, validate_model
from pydantic.fields import Shape

from .json_tools import pretty_lenient_json

JSON_CONTENT_TYPE = 'application/json'
IP_HEADER = 'X-Forwarded-For'
PROTO_HEADER = 'X-Forwarded-Proto'
URI_NOT_ALLOWED = re.compile(r'[^a-zA-Z0-9_\-/.]')
REMOVE_PORT = re.compile(r':\d{2,}$')
PydanticModel = TypeVar('PydanticModel', bound=BaseModel)

__all__ = (
    'raw_json_response',
    'json_response',
    'parse_request_json',
    'parse_request_json_ignore_missing',
    'parse_request_query',
    'get_ip',
    'request_root',
    'JsonErrors',
    'encrypt_json',
    'decrypt_json',
    'get_offset',
    'slugify',
    'remove_port',
    'RequestError',
)


def raw_json_response(json_str: Union[str, bytes, None], status_: int = 200):
    if isinstance(json_str, str):
        body = json_str.encode()
    elif isinstance(json_str, bytes):
        body = json_str
    elif json_str is None:
        body = b'null'
    else:
        raise TypeError(f'json_str must be bytes or str, not "{type(json_str)}')
    return Response(body=body + b'\n', status=status_, content_type=JSON_CONTENT_TYPE)


def json_response(*, status_=200, list_=None, headers_=None, **data):
    return Response(
        body=json.dumps(data if list_ is None else list_).encode() + b'\n',
        status=status_,
        content_type=JSON_CONTENT_TYPE,
        headers=headers_,
    )


async def parse_request_json(request, model: Type[PydanticModel], *, headers=None) -> PydanticModel:
    error_details = None
    try:
        data = await request.json()
    except ValueError:
        error_msg = 'Invalid JSON'
    else:
        try:
            return model.parse_obj(data)
        except ValidationError as e:
            error_msg = 'Invalid Data'
            error_details = e.errors()

    raise JsonErrors.HTTPBadRequest(message=error_msg, details=error_details, headers=headers)


def parse_request_query(request, model: Type[PydanticModel], *, headers=None) -> PydanticModel:
    data = {}
    for k in request.query:
        v = request.query.getall(k)
        f = model.__fields__.get(k)
        if len(v) > 1 or f and f.shape != Shape.SINGLETON:
            data[k] = v
        else:
            data[k] = v[0]

    try:
        return model(**data)
    except ValidationError as e:
        raise JsonErrors.HTTPBadRequest(message='Invalid query data', details=e.errors(), headers=headers)


async def parse_request_json_ignore_missing(
    request, model: Type[PydanticModel], *, headers=None
) -> Tuple[PydanticModel, dict]:
    try:
        raw_data = await request.json()
    except ValueError:
        raise JsonErrors.HTTPBadRequest(message='Invalid JSON', headers=headers)
    if not isinstance(raw_data, dict):
        raise JsonErrors.HTTPBadRequest(message='data not a dictionary', headers=headers)

    data, e = validate_model(model, raw_data, raise_exc=False)
    if e:
        errors = [e for e in e.errors() if not (e['type'] == 'value_error.missing' and len(e['loc']) == 1)]
        if errors:
            raise JsonErrors.HTTPBadRequest(message='Invalid Data', details=errors, headers=headers)

    return model.construct(**data), raw_data


def get_ip(request):
    ips = request.headers.get(IP_HEADER)
    if ips:
        return ips.split(',', 1)[0].strip(' ')
    else:
        return request.remote


def request_root(request):
    # request.url.scheme doesn't work as https is already terminated
    scheme = request.headers.get(PROTO_HEADER) or 'http'
    return f'{scheme}://{request.host}'


class JsonErrors:
    class _HTTPExceptionJson(HTTPException):
        custom_reason = None

        def __init__(self, message, *, details=None, headers=None):
            self.message = message
            self.details = details
            data = {'message': message}
            if details:
                data['details'] = details
            super().__init__(
                text=pretty_lenient_json(data),
                content_type=JSON_CONTENT_TYPE,
                headers=headers,
                reason=self.custom_reason,
            )

        def __repr__(self) -> str:
            return f'{super().__repr__()}, {self.status}: {self.message}'

        def __str__(self) -> str:
            return repr(self)

    class HTTPAccepted(_HTTPExceptionJson):
        status_code = 202

    class HTTPBadRequest(_HTTPExceptionJson):
        status_code = 400

    class HTTPUnauthorized(_HTTPExceptionJson):
        status_code = 401

    class HTTPPaymentRequired(_HTTPExceptionJson):
        status_code = 402

    class HTTPForbidden(_HTTPExceptionJson):
        status_code = 403

    class HTTPNotFound(_HTTPExceptionJson):
        status_code = 404

    class HTTPMethodNotAllowed(_HTTPExceptionJson):
        status_code = 405

        def __init__(self, message, allowed_methods, *, headers=None):
            headers = headers or {}
            headers.setdefault('Allow', ','.join(allowed_methods))
            super().__init__(message, details={'allowed_methods': allowed_methods}, headers=headers)

    class HTTPConflict(_HTTPExceptionJson):
        status_code = 409

    class HTTP470(_HTTPExceptionJson):
        status_code = 470
        custom_reason = 'Invalid user input'


def encrypt_json(app, data: Any) -> str:
    return app['auth_fernet'].encrypt(json.dumps(data).encode()).decode()


def decrypt_json(app, token: bytes, *, ttl: int = None, headers=None) -> Any:
    from cryptography.fernet import InvalidToken

    try:
        return json.loads(app['auth_fernet'].decrypt(token, ttl=ttl).decode())
    except InvalidToken:
        raise JsonErrors.HTTPBadRequest(message='invalid token', headers=headers)


def get_offset(request, paginate_by=100):
    page = request.query.get('page')
    if not page:
        return 0

    try:
        p = int(page)
        if p < 1:
            raise ValueError()
    except ValueError:
        raise JsonErrors.HTTPBadRequest(f"invalid page '{page}'")
    else:
        return (p - 1) * paginate_by


def slugify(title):
    name = title.replace(' ', '-').lower()
    name = URI_NOT_ALLOWED.sub('', name)
    name = re.sub('-{2,}', '-', name)
    return name.strip('_-')


def remove_port(host):
    return REMOVE_PORT.sub('', host)


class RequestError(RuntimeError):
    def __init__(self, status, url, *, text: str = None):
        self.status = status
        self.url = url
        self.text = text

    def __str__(self):
        return f'response {self.status} from "{self.url}"' + (f':\n{self.text[:400]}' if self.text else '')

    def json(self):
        return json.loads(self.text)

    def extra(self):
        return self.text
