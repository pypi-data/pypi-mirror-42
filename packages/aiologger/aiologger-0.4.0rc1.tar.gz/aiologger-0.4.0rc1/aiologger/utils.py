from typing import Callable


class classproperty:
    def __init__(self, func):
        self._func = func

    def __get__(self, obj, owner):
        return self._func(owner)


class CallableWrapper:
    def __init__(self, func: Callable) -> None:
        self.func = func

    def __call__(self, *args, **kwargs):
        return self.func(*args, **kwargs)
