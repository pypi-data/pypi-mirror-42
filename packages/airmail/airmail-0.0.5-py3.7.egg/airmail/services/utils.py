import os, sys
from pathlib import Path
import click
import yaml

def err_log(msg, *args):
    if args:
        msg %= args
    click.echo(click.style('Error: ', fg='red', bold=True) + msg)

# Read a yaml file given a file path. Shou
def read_yml(filepath):
    target_file = Path(filepath)

    if target_file.is_file():
        with open(target_file) as yml_file:
            return yaml.load(yml_file)
    else:
        click.echo(click.style('Error: ', fg='red', bold=True) +
            'Could not read the .yml at ' +
            click.style(filepath, bold=True))
        sys.exit()

# Determines the path to files inside the
# project. (i.e. the `data` directory)
def determine_project_path():
    """Borrowed from wxglade.py"""
    try:
        root = __file__
        if os.path.islink (root):
            root = os.path.realpath (root)
        return os.path.dirname (os.path.abspath (root))
    except:
        print("I'm sorry, but something is wrong.")
        print("There is no __file__ variable. Please contact the author.")
        sys.exit ()
