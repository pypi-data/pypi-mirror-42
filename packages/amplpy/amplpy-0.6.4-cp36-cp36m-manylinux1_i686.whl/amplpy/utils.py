# -*- coding: utf-8 -*-
from __future__ import print_function, absolute_import, division
from builtins import map, range, object, zip, sorted
from past.builtins import basestring, unicode
from sys import version_info

from . import amplpython
from .base import BaseClass


def lock_and_call(call, lock):
    lock.acquire()
    try:
        result = call()
    except Exception:
        lock.release()
        raise
    else:
        lock.release()
        return result


class Tuple(BaseClass):
    def __init__(self, values=None, **kwargs):
        if values is not None:
            values = Utils.convToList(values)
            va = Utils.toVariantArray(values)
            self._impl = amplpython.Tuple.Factory(va, len(values))
        else:
            self._impl = kwargs.get('_impl', None)

    def toTuple(self):
        return tuple(
            Utils.castVariant(self._impl.getIndex(i))
            for i in range(self._impl.size())
        )

    def toPyObj(self):
        size = self._impl.size()
        if size == 1:
            return Utils.castVariant(self._impl.getIndex(0))
        else:
            return self.toTuple()

    @classmethod
    def fromTupleRef(cls, tupleRef):
        return cls(_impl=amplpython.Tuple.Factory(tupleRef))


class Utils(object):
    @staticmethod
    def toVariantArray(lst):
        va = amplpython.VariantArray(len(lst))
        for i in range(len(lst)):
            if isinstance(lst[i], unicode):
                va[i] = amplpython.Variant(str(lst[i]))
                # FIXME: This is just a workaround for issue amplapi#332
                # The real fix requires a new release of amplapi
            else:
                va[i] = amplpython.Variant(lst[i])
        return va

    @staticmethod
    def toTupleArray(lst):
        ta = amplpython.TupleArray(len(lst))
        for i in range(len(lst)):
            ta[i] = Tuple(lst[i])._impl
        return ta

    @staticmethod
    def castVariant(variant):
        if variant.type() == amplpython.NUMERIC:
            return variant.dbl()
        elif variant.type() == amplpython.STRING:
            return variant.str()

    @staticmethod
    def castStringArray(sarray):
        return [sarray.getIndex(i) for i in range(sarray.size())]

    @staticmethod
    def castVariantRef(variantref):
        return Utils.castVariant(amplpython.Variant(variantref))

    @staticmethod
    def convToList(value):
        if isinstance(value, list):
            return value
        elif isinstance(value, tuple):
            return list(value)
        else:
            return [value]


def multidict(d):
    try:
        ncols = min(len(d[k]) for k in d)
    except TypeError:
        raise TypeError
    return [list(d.keys())] + [{k: d[k][i] for k in d} for i in range(ncols)]
