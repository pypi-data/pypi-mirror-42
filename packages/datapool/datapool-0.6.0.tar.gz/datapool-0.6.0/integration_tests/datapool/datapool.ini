[db]
connection_string = sqlite+pysqlite:////Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz/.simple.db

[backup_landing_zone]
folder = 

[r]
extension = .r
executable = /usr/local/bin/R

[matlab]
extension = .m
executable = /Applications/MATLAB_R2016a.app/bin/matlab

[julia]
extension = .jl
version = 0.5.2
executable = /usr/local/bin/julia

[python]
extension = .py
executable = /Users/uweschmitt/Projects/eawag_datapool/venv/bin/python3

[logging]
config_file = ./logging_config.yaml

[log_receiver]
port = 5559

[worker]
port = 5555
count = 5

[server]
pid_file = /Users/uweschmitt/.datapool/datapool.pid

[landing_zone]
folder = /Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz

