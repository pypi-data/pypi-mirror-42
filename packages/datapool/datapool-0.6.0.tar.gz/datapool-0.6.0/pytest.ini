[pytest]
addopts = -rs --full-trace
testpaths = tests
