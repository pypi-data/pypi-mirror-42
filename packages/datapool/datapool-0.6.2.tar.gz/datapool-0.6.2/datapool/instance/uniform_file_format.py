# encoding: utf-8
from __future__ import print_function, division, absolute_import

import csv
import datetime
import io
import os

from datapool.utils import iter_to_list


def _check_header(header):

    header = list(header)

    if "site" in header:
        header.remove("site")
    else:
        for n in "xyz":
            if n not in header:
                raise ValueError(
                    "header {!r} must have either 'site' or 'x', 'y' and 'z'".format(
                        header
                    )
                )
            header.remove(n)

    header = set(header)
    fix_column_names = set(("timestamp", "parameter", "value"))
    if header != fix_column_names:
        missing = ["'{}'".format(name) for name in fix_column_names - header]
        invalid = ["'{}'".format(name) for name in header - fix_column_names]
        msg = ""
        if missing:
            msg += "missing: {}".format(", ".join(sorted(missing)))
        if invalid:
            msg += "invalid: {}".format(", ".join(sorted(invalid)))
        raise ValueError("invalid header ({})".format(msg))


@iter_to_list
def read_from_file(path):
    if not os.path.exists(path):
        raise ValueError("{} does not exist".format(path))

    with open(path, "r", encoding="ascii") as fh:
        yield from _read_from_fh(fh)


@iter_to_list
def read_from_string(data):
    fh = io.StringIO(data)
    yield from _read_from_fh(fh)


def can_be_converted_to_float(value):
    try:
        value = float(value)
    except ValueError:
        return False
    return True


def parse_timestamp(timestamp):
    return datetime.datetime.strptime(timestamp, "%Y-%m-%d %H:%M:%S")


def _read_from_fh(fh):
    reader = csv.reader(fh, delimiter=";")
    header = [f.strip() for f in next(reader)]
    _check_header(header)
    for i, row in enumerate(reader):
        if len(row) < len(header):
            # i + 1, we skipped header
            raise ValueError("row {} ({!r}) is incomplete".format(i + 1, row))
        if len(row) > len(header):
            # i + 1, we skipped header
            raise ValueError("row {} ({!r}) is to long".format(i + 1, row))

        row = (cell.strip() for cell in row)
        row_dict = dict(zip(header, row))
        yield row_dict


@iter_to_list
def check_rows(row_dicts):

    if not row_dicts:
        return

    for i, row_dict in enumerate(row_dicts):

        value = row_dict["value"]
        if not can_be_converted_to_float(value):
            yield "row {}: value '{}' is not a float".format(i, value)
        else:
            row_dict["value"] = float(value)

        timestamp = row_dict["timestamp"]
        try:
            row_dict["timestamp"] = parse_timestamp(timestamp)
        except ValueError:
            yield "row {}: timstamp '{}' either has wrong format or is invalid".format(
                i, timestamp
            )

        has_xyz = all(name in row_dict and row_dict[name] is not None for name in "xyz")
        not_empty = ("parameter",)
        if not has_xyz:
            not_empty += ("site",)

        for name in not_empty:
            if row_dict[name].strip() == "":
                yield "row {}: field {} is empty".format(i, name)


@iter_to_list
def to_signals(row_dicts):

    for row in row_dicts:
        for f in "xyz":
            if f in row:
                row["coord_" + f] = row[f]
                del row[f]

    from .domain_objects import Signal

    return list(map(Signal, row_dicts))
