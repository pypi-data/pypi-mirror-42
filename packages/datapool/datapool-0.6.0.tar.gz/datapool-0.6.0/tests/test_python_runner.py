# encoding: utf-8
from __future__ import print_function, division, absolute_import

from datapool.python_runner import PythonRunner
from datapool.instance.config_handling import guess_config
from datapool.utils import find_executable

import pytest

has_python = find_executable("python") is not None

# we skip this test file if matlab is not installed on the test system:
# see http://doc.pytest.org/en/latest/skipping.html#skip-all-test-functions-of-a-class-or-module
pytestmark = pytest.mark.skipif(
    not has_python, reason="no python interpreter found on your system"
)


@pytest.fixture
def python_exe(regtest):
    config, messages = guess_config({})
    assert config.python.executable != "", str(config)
    return config.python.executable


def test_python(python_exe, setup_logger):
    r = PythonRunner(python_exe)
    r.start_interpreter()
    assert r.is_alive()


def test_python_exception(python_exe, regtest_logger):
    r = PythonRunner(python_exe)
    r.start_interpreter()
    with regtest_logger():
        assert r.run_command("1/0") == 1
    assert r.is_alive()
