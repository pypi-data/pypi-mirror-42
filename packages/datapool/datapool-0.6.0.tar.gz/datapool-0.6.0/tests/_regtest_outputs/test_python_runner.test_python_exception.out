ERROR    eawag.datapool  python: Traceback (most recent call last):
ERROR    eawag.datapool  python:   File "<stdin>", line 2, in <module>
ERROR    eawag.datapool  python: ZeroDivisionError: division by zero
ERROR    eawag.datapool  python: 
